File: Creational/VirtualContstructor-Factory/javaScript/classes.js
export class Product {
  constructor() {
    if (this.constructor === Product) {
      throw new Error("Abstract class can not be instantiated.");
    }
  }

  doStuff() {
    throw new Error("Cannot call abstract method");
  }
}

export class Creator {
  createProduct() {
    throw new Error("Cannot call abstract method");
  }

  someOperation() {
    const product = this.createProduct();
    if (!(product instanceof Product)) {
      throw new Error("product has to be type of Product");
    }
    product.doStuff();
  }
}

File: Creational/VirtualContstructor-Factory/python/classes.py
from abc import ABC, abstractmethod

class Product(ABC):

    @abstractmethod
    def do_stuff(self):
        pass


class Creator(ABC):

    # this is the factory method
    @abstractmethod
    def create_product(self) -> Product:
        pass

    def some_operation(self):
        p: Product = self.create_product()
        p.do_stuff()
